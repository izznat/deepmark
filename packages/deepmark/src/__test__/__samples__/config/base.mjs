/** @type {import("../../../config").UserConfig} */
export default {
	sourceLanguage: 'en',
	outputLanguages: ['zh', 'ja'],
	directories: [
		['i18n/$langcode$', 'i18n/$langcode$'],
		['docs', 'i18n/$langcode$/docs'],
		['blog', 'i18n/$langcode$/blog']
	],
	cwd: '../../../../example'
};
