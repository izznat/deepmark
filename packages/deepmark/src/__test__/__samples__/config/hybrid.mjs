import base from './base.mjs';

/** @type {import("../../../config").UserConfig} */
export default {
	...base,
	files: {
		include: ['docs/intro.md', 'docs/tutorial-basics/markdown-features.mdx', 'i18n/en/code.json']
	},
	jsonOrYamlProperties: {
		include: ['message', 'description']
	}
};
