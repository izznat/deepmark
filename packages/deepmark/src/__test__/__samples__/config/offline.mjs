import base from './base.mjs';

/** @type {import("../../../config").UserConfig} */
export default {
	...base,
	jsonOrYamlProperties: {
		include: ['message', 'description']
	}
};
