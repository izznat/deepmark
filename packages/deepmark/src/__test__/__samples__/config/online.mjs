import base from './base.mjs';

/** @type {import("../../../config").UserConfig} */
export default {
	...base,
	files: {
		include: ['docs/tutorial-basics/markdown-features.mdx', 'i18n/en/code.json']
	},
	jsonOrYamlProperties: {
		include: ['message', 'description']
	}
};
