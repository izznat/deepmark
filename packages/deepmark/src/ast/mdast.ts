import type {
	Root as MdRoot,
	Blockquote as MdBlockquote,
	Break as MdBreak,
	Code as MdCode,
	Definition as MdDefinition,
	Delete as MdDelete,
	Emphasis as MdEmphasis,
	Footnote as MdFootnote,
	FootnoteDefinition as MdFootnoteDefinition,
	FootnoteReference as MdFootnoteReference,
	HTML as MdHTML,
	Heading as MdHeading,
	Image as MdImage,
	ImageReference as MdImageReference,
	InlineCode as MdInlineCode,
	Link as MdLink,
	LinkReference as MdLinkReference,
	List as MdList,
	ListItem as MdListItem,
	Paragraph as MdParagraph,
	Strong as MdStrong,
	Table as MdTable,
	TableCell as MdTableCell,
	TableRow as MdTableRow,
	Text as MdText,
	ThematicBreak as MdThematicBreak,
	YAML as MdYaml
} from 'mdast';

import type {
	MdxFlowExpression,
	MdxJsxAttribute,
	MdxJsxAttributeValueExpression,
	MdxJsxExpressionAttribute,
	MdxJsxFlowElement,
	MdxJsxTextElement,
	MdxTextExpression,
	MdxjsEsm
} from 'mdast-util-mdx';

import type { UnNode } from './unist.js';

import { fromMarkdown } from 'mdast-util-from-markdown';
import { frontmatterFromMarkdown, frontmatterToMarkdown } from 'mdast-util-frontmatter';
import { htmlCommentFromMarkdown, htmlCommentToMarkdown } from 'mdast-util-html-comment';
import { mdxFromMarkdown, mdxToMarkdown } from 'mdast-util-mdx';
import { toMarkdown } from 'mdast-util-to-markdown';
import { frontmatter } from 'micromark-extension-frontmatter';
import { htmlComment } from 'micromark-extension-html-comment';
import { mdxjs } from 'micromark-extension-mdxjs';

declare module 'mdast' {
	export interface PhrasingContentMap extends StaticPhrasingContentMap {
		mdxJsxFlowElement: MdxJsxFlowElement;
		mdxJsxTextElement: MdxJsxTextElement;
		mdxFlowExpression: MdxFlowExpression;
		mdxTextExpression: MdxTextExpression;
	}
}

export function mdNodeIs<T extends MdNodeType>(
	node: UnNode | undefined,
	type: T
): node is T extends MdRoot['type']
	? MdRoot
	: T extends MdBlockquote['type']
	? MdBlockquote
	: T extends MdBreak['type']
	? MdBreak
	: T extends MdCode['type']
	? MdCode
	: T extends MdDefinition['type']
	? MdDefinition
	: T extends MdDelete['type']
	? MdDelete
	: T extends MdEmphasis['type']
	? MdEmphasis
	: T extends MdFootnote['type']
	? MdFootnote
	: T extends MdFootnoteDefinition['type']
	? MdFootnoteDefinition
	: T extends MdFootnoteReference['type']
	? MdFootnoteReference
	: T extends MdHTML['type']
	? MdHTML
	: T extends MdHeading['type']
	? MdHeading
	: T extends MdImage['type']
	? MdImage
	: T extends MdImageReference['type']
	? MdImageReference
	: T extends MdInlineCode['type']
	? MdInlineCode
	: T extends MdLink['type']
	? MdLink
	: T extends MdLinkReference['type']
	? MdLinkReference
	: T extends MdList['type']
	? MdList
	: T extends MdListItem['type']
	? MdListItem
	: T extends MdParagraph['type']
	? MdParagraph
	: T extends MdStrong['type']
	? MdStrong
	: T extends MdTable['type']
	? MdTable
	: T extends MdTableCell['type']
	? MdTableCell
	: T extends MdTableRow['type']
	? MdTableRow
	: T extends MdText['type']
	? MdText
	: T extends MdThematicBreak['type']
	? MdThematicBreak
	: T extends MdYaml
	? MdYaml
	: T extends MdxFlowExpression['type']
	? MdxFlowExpression
	: T extends MdxJsxAttribute['type']
	? MdxJsxAttribute
	: T extends MdxJsxAttributeValueExpression['type']
	? MdxJsxAttributeValueExpression
	: T extends MdxJsxExpressionAttribute['type']
	? MdxJsxExpressionAttribute
	: T extends MdxJsxFlowElement['type']
	? MdxJsxFlowElement
	: T extends MdxJsxTextElement['type']
	? MdxJsxTextElement
	: T extends MdxTextExpression['type']
	? MdxTextExpression
	: MdxjsEsm {
	return node ? node.type === type : false;
}

export function mdNodeIsJsxElement(node: UnNode): node is MdxJsxFlowElement | MdxJsxTextElement {
	return mdNodeIs(node, 'mdxJsxFlowElement') || mdNodeIs(node, 'mdxJsxTextElement');
}

/**
 * Get MDX flavored `mdast`.
 */
export function getMdast(markdown: string): MdRoot {
	return fromMarkdown(markdown, {
		extensions: [frontmatter('yaml'), mdxjs(), htmlComment()],
		mdastExtensions: [frontmatterFromMarkdown('yaml'), mdxFromMarkdown(), htmlCommentFromMarkdown()]
	});
}

export function getMarkdown(mdast: MdRoot): string {
	return toMarkdown(mdast, {
		extensions: [frontmatterToMarkdown('yaml'), mdxToMarkdown(), htmlCommentToMarkdown()],
		join: [
			(__, _, parent) => {
				if (mdNodeIsJsxElement(parent)) {
					return 0;
				}

				return 1;
			}
		]
	});
}

/**
 * ============================================================
 */

export type MdNodeType =
	| MdRoot['type']
	| MdBlockquote['type']
	| MdBreak['type']
	| MdCode['type']
	| MdDefinition['type']
	| MdDelete['type']
	| MdEmphasis['type']
	| MdFootnote['type']
	| MdFootnoteDefinition['type']
	| MdFootnoteReference['type']
	| MdHTML['type']
	| MdHeading['type']
	| MdImage['type']
	| MdImageReference['type']
	| MdInlineCode['type']
	| MdLink['type']
	| MdLinkReference['type']
	| MdList['type']
	| MdListItem['type']
	| MdParagraph['type']
	| MdStrong['type']
	| MdTable['type']
	| MdTableCell['type']
	| MdTableRow['type']
	| MdText['type']
	| MdThematicBreak['type']
	| MdYaml['type']
	| MdxFlowExpression['type']
	| MdxJsxAttribute['type']
	| MdxJsxAttributeValueExpression['type']
	| MdxJsxExpressionAttribute['type']
	| MdxJsxFlowElement['type']
	| MdxJsxTextElement['type']
	| MdxTextExpression['type']
	| MdxjsEsm['type'];

export type {
	MdRoot,
	MdBlockquote,
	MdBreak,
	MdCode,
	MdDefinition,
	MdDelete,
	MdEmphasis,
	MdFootnote,
	MdFootnoteDefinition,
	MdFootnoteReference,
	MdHTML,
	MdHeading,
	MdImage,
	MdImageReference,
	MdInlineCode,
	MdLink,
	MdLinkReference,
	MdList,
	MdListItem,
	MdParagraph,
	MdStrong,
	MdTable,
	MdTableCell,
	MdTableRow,
	MdText,
	MdThematicBreak,
	MdYaml
};

export type {
	MdxFlowExpression,
	MdxJsxAttribute,
	MdxJsxAttributeValueExpression,
	MdxJsxExpressionAttribute,
	MdxJsxFlowElement,
	MdxJsxTextElement,
	MdxTextExpression,
	MdxjsEsm
};
