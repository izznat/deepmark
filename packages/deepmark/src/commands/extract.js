import fg from 'fast-glob';
import fs from 'fs-extra';
import np from 'path';
import { get_path_tail, get_string_array, is_directory_exist, resolve_path } from '../utils.js';

/**
 * @param { import("../types.js").Config } config
 * @returns { import("../types.js").CommandHandler }
 */
export function extract({ sourceLanguage, directories: { extraction }, frontmatter, components }) {
	return async () => {
		const source_dirs = get_string_array(extraction.source)
		const output_dirs = get_string_array(extraction.output)

		for (let i = 0; i < source_dirs.length; i++) {
			const source_dir = source_dirs[i].replace('%language%', sourceLanguage);
			if (!is_directory_exist(source_dir)) continue;

			const source_pattern = np.join(source_dir, '**/*.{md,mdx}');
			const source_paths = (await fg(source_pattern)).map((path) => resolve_path(path));

			const output_dir = output_dirs[i].replace('%language%', sourceLanguage);

			for (const source_path of source_paths) {
				const output_path = resolve_path(output_dir, get_path_tail(output_dir, source_path));

				/** @type { string[] } */
				const strings = [];

				await fs.outputFile(output_path, JSON.stringify(strings));
			}
		}
	};
}
