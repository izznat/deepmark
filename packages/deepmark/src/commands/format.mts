import type { Config } from '$types';

import fg from 'fast-glob';
import fs from 'fs-extra';
import np from 'path';
import { format } from '../features/index.mjs';
import { get_string_array, is_directory_exist, resolve_path } from '../utilities/index.mjs';

export function create_format_handler(config: Config) {
	const {
		directories: { sources }
	} = config;

	return async () => {
		const source_dirs = get_string_array(sources);

		for (const source_dir of source_dirs) {
			if (!is_directory_exist(source_dir)) continue;

			const source_pattern = np.join(source_dir, '**/*.{md,mdx}');
			const source_paths = (await fg(source_pattern)).map((path) => resolve_path(path));

			for (const source_path of source_paths) {
				const output_path = source_path;

				const markdown = await fs.readFile(source_path, { encoding: 'utf-8' });
				const formatted = await format(markdown);

				await fs.writeFile(output_path, formatted);
			}
		}
	};
}
