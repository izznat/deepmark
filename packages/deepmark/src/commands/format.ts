import type { FormatOptions } from '../features/index.js';
import type { Config } from '../types/index.js';

import fg from 'fast-glob';
import fs from 'fs-extra';
import { format_markdown } from '../features/index.js';
import { get_string_array, is_directory_exist, resolve_path } from '../utilities/index.js';

export function create_format_handler(config: Config) {
	const {
		directories: { sources }
	} = config;

	return async (options: FormatOptions) => {
		const source_dirs = get_string_array(sources);

		for (const source_dir of source_dirs) {
			if (!is_directory_exist(source_dir)) continue;

			const source_pattern = source_dir.replace(/\/$/, '') + '/**/*.{md,mdx}';
			const source_paths = (await fg(source_pattern)).map((path) => resolve_path(path));

			for (const source_path of source_paths) {
				const output_path = source_path;

				const markdown = await fs.readFile(source_path, { encoding: 'utf-8' });
				const formatted = await format_markdown(markdown, options);

				await fs.writeFile(output_path, formatted);
			}
		}
	};
}
