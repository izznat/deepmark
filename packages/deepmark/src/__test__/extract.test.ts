import fs from 'fs-extra';
import np from 'node:path';
import { config } from 'node:process';
import { describe } from 'vitest';
import { getMdast } from '../ast/mdast';
import { Config, resolveConfig, UserConfig } from '../config';
import { extractJsonOrYamlStrings, extractMdastStrings } from '../extract';
import { format } from '../format';

const baseConfig: UserConfig = {
	sourceLanguage: 'en',
	outputLanguages: ['zh'],
	directories: [['', '']],
	cwd: 'src/__test__/__samples__'
};

async function extract(
	path: string,
	config: Config,
	from: 'markdown' | 'json' | 'yaml' = 'markdown'
) {
	const resolvedPath = np.resolve(config.cwd, path);
	const source = await fs.readFile(resolvedPath, { encoding: 'utf-8' });

	if (from === 'markdown')
		return extractMdastStrings({ mdast: getMdast(await format(source)), config });

	return extractJsonOrYamlStrings({ source, type: from, config });
}

describe('extract frontmatter field string values', (test) => {
	test('ignore empty frontmatter', async ({ expect }) => {
		const strings = await extract('frontmatter/empty.md', resolveConfig(baseConfig));

		expect(strings.length).toBe(0);
	});

	test('filter frontmatter fields based on configuration', async ({ expect }) => {
		const strings = await extract(
			'frontmatter/index.md',
			resolveConfig({
				...baseConfig,
				frontmatterFields: {
					include: ['title', 'tags', 'description'],
					exclude: ['description']
				}
			})
		);

		expect(strings).toMatchSnapshot();
	});
});

describe('extract jsx children and attribute string values', (test) => {
	test('recursively extract strings from nested jsx components', async ({ expect }) => {
		const strings = await extract(
			'jsx/nested.mdx',
			resolveConfig({
				...baseConfig,
				jsxComponents: {
					include: {
						Block: { children: true, attributes: [] }
					}
				}
			})
		);

		expect(strings).toMatchSnapshot();
	});

	test('recursively extract strings from html elements and jsx components inside attributes', async ({
		expect
	}) => {
		const strings = await extract(
			'jsx/jsx-in-prop.mdx',
			resolveConfig({
				...baseConfig,
				jsxComponents: {
					include: {
						Card: {
							children: true,
							attributes: ['header']
						},
						List: {
							children: false,
							attributes: ['items']
						}
					}
				}
			})
		);

		expect(strings).toMatchSnapshot();
	});

	test('ignore some HTML elements by default', async ({ expect }) => {
		const strings = await extract('jsx/code-and-pre.mdx', resolveConfig(baseConfig));

		expect(strings).toMatchSnapshot();
	});
});

describe('extract strings from JSON based on configuration', (test) => {
	test('do not extract any string if no property name is included in the config', async ({
		expect
	}) => {
		const strings = await extract('json/navbar.json', resolveConfig(baseConfig), 'json');

		expect(strings.length).toBe(0);
	});

	test('filter properties based on the config', async ({ expect }) => {
		const strings = await extract(
			'json/navbar.json',
			resolveConfig({
				...baseConfig,
				jsonOrYamlProperties: {
					include: ['message', 'description'],
					exclude: ['message']
				}
			}),
			'json'
		);

		expect(strings).toMatchSnapshot();
	});
});

describe('extract strings from yaml based on configuration', (test) => {
	test('do not extract any string if no property name is included in the config', async ({
		expect
	}) => {
		const strings = await extract('yaml/authors.yml', resolveConfig(baseConfig), 'yaml');

		expect(strings.length).toBe(0);
	});

	test('filter properties based on the config', async ({ expect }) => {
		const strings = await extract(
			'yaml/authors.yml',
			resolveConfig({
				...baseConfig,
				jsonOrYamlProperties: {
					include: ['name', 'title'],
					exclude: ['name']
				}
			}),
			'yaml'
		);

		expect(strings).toMatchSnapshot();
	});
});
