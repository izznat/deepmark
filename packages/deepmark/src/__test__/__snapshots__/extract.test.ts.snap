// Vitest Snapshot v1

exports[`extract frontmatter field string values > filter frontmatter fields based on configuration 1`] = `
[
  "A Short Title",
  "tagone",
  "tagtwo",
]
`;

exports[`extract jsx children and attribute string values > ignore some HTML elements by default 1`] = `
[
  "This is a text. ",
]
`;

exports[`extract jsx children and attribute string values > recursively extract strings from html elements and jsx components inside attributes 1`] = `
[
  "This is a text inside a custom component.",
  "This is a text inside a jsx prop.",
  "This is the text of jsx item one. ",
  "This the nested text of jsx item one.",
  "A short title inside title attribute inside HTML element inside an attribute.",
  "This is the text of jsx item two. ",
  "This the nested text of jsx item two.",
]
`;

exports[`extract jsx children and attribute string values > recursively extract strings from nested jsx components 1`] = `
[
  "This is a paragraph.",
  "This is a span.",
  "This is a text inside a custom component.",
]
`;

exports[`extract strings from JSON based on configuration > filter properties based on the config 1`] = `
[
  "The title in the navbar",
  "Navbar item with label Tutorial",
  "Navbar item with label Blog",
  "Navbar item with label GitHub",
]
`;

exports[`extract strings from yaml based on configuration > filter properties based on the config 1`] = `
[
  "Maintainer of Docusaurus",
  "Front End Engineer @ Facebook",
  "Docusaurus maintainer",
]
`;
